### PR TITLE
Add `napari_experimental_provide_theme` hook specification

### PR DESCRIPTION
**Pull request by lukasz-migas**
This PR is an offshot of #2900 that will enable easy addition of icons, stylesheets and themes. Since there is still a bit more to do there, the `napari_provide_theme` spec was separated as it already provides the relevant functionality.

# Description
- Add a way to provide `theme` that controls the look of napari
- Add new `Theme` class that is an evented model
- Changes to the active theme are now immediately visible in the app

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# References
#2900 - original PR
#2994  - evented dict

# How has this been tested?
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

evented-theme

### CODE REVIEW

File: examples/new_theme.py
Comment (sofroniewn) [inline examples/new_theme.py:24]: Love this!

File: napari/utils/theme.py
Comment (sofroniewn) [inline napari/utils/theme.py:24]: Very nice

**Review by sofroniewn — COMMENTED** [review]
This looks great! I think we just need to fix the failing tests - looks like there might still be some event connection left over that triggers an update to a deleted object - see 

```python-traceback
==================================== ERRORS ====================================
  ___________________ ERROR at teardown of test_changing_theme ___________________
  
  qtbot = <pytestqt.qtbot.QtBot object at 0x7ff551c75cd0>
  request = <SubRequest 'make_napari_viewer' for <Function test_changing_theme>>
  napari_plugin_manager = <napari.plugins._plugin_manager.NapariPluginManager object at 0x7ff551c75ee0>
  
      @pytest.fixture
      def make_napari_viewer(
          qtbot, request: 'FixtureRequest', napari_plugin_manager
      ):
          """A fixture function that creates a napari viewer for use in testing.
      
          Use this fixture as a function in your tests:
      
              viewer = make_napari_viewer()
      
          It accepts all the same arguments as napari.Viewer, plus the following
          test-related paramaters:
      
          ViewerClass : Type[napari.Viewer], optional
              Override the viewer class being used.  By default, will
              use napari.viewer.Viewer
          strict_qt : bool or str, optional
              If True, a check will be performed after test cleanup to make sure that
              no top level widgets were created and *not* cleaned up during the
              test.  If the string "raise" is provided, an AssertionError will be
              raised.  Otherwise a warning is emitted.
              By default, this is False unless the test is being performed within
              the napari package.
              This can be made globally true by setting the 'NAPARI_STRICT_QT'
              environment variable.
          block_plugin_discovery : bool, optional
              Block discovery of non-builtin plugins.  Note: plugins can still be
              manually registered by using the 'napari_plugin_manager' fixture and
              the `napari_plugin_manager.register()` method. By default, True.
      
          Examples
          --------
          >>> def test_adding_shapes(make_napari_viewer):
          ...     viewer = make_napari_viewer()
          ...     viewer.add_shapes()
          ...     assert len(viewer.layers) == 1
      
          >>> def test_something_with_plugins(make_napari_viewer):
          ...     viewer = make_napari_viewer(block_plugin_discovery=False)
      
          >>> def test_something_with_strict_qt_tests(make_napari_viewer):
  napari/utils/events/evented_model.py:155: in __setattr__
      getattr(self.events, name)(value=after)  # emit event
  napari/utils/events/event.py:575: in __call__
      self._invoke_callback(cb, event)
  napari/utils/events/event.py:598: in _invoke_callback
      _handle_exception(
  napari/utils/events/event.py:596: in _invoke_callback
      cb(event)
  napari/_qt/qt_main_window.py:1561: in _update_theme
      self.qt_viewer.viewer.theme = value
  napari/components/viewer_model.py:323: in __setattr__
      return super().__setattr__(name, value)
  napari/utils/events/evented_model.py:155: in __setattr__
      getattr(self.events, name)(value=after)  # emit event
  napari/utils/events/event.py:575: in __call__
      self._invoke_callback(cb, event)
  napari/utils/events/event.py:598: in _invoke_callback
      _handle_exception(
  napari/utils/events/event.py:596: in _invoke_callback
      cb(event)
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  self = <[RuntimeError('Internal C++ object (QtAboutKeyBindings) already deleted.') raised in repr()] QtAboutKeyBindings object at 0x7ff5843c5fc0>
  event = <Event blocked=False handled=False native=None source=None sources=[] type='theme'>
  
      def update_active_layer(self, event=None):
          """Update the active layer and display key bindings for that layer type.
      
          Parameters
          ----------
          event : napari.utils.event.Event, optional
              The napari event that triggered this method, by default None.
          """
          theme = get_theme(self.viewer.theme)
          col = theme['secondary']
      
          # Add class and instance viewer key bindings
          active_shortcuts = action_manager._get_active_shortcuts(
              self.key_map_handler.active_keymap
          )
          text = get_key_bindings_summary(active_shortcuts, col=col)
      
          # Update layer speficic key bindings if all active are displayed
          self.key_bindings_strs[self.ALL_ACTIVE_KEYBINDINGS] = text
  >       if self.layerTypeComboBox.currentText() == self.ALL_ACTIVE_KEYBINDINGS:
  E       RuntimeError: Internal C++ object (PySide2.QtWidgets.QComboBox) already deleted.
  
  napari/_qt/dialogs/qt_about_key_bindings.py:151: RuntimeError
  =================================== FAILURES ===================================
```

**Comment by lukasz-migas** [comment]
Yeah I think you might be right. Going to add a couple of tests to cover the addition of new plugins and try to figure out where the failure is coming from.

File: napari/utils/theme.py
Comment (tlambert03) [inline napari/utils/theme.py]: `folder` seems pretty antiquated at this point.  This started out as the literal folder that the icons were in, but we generally use it as the name of the theme right?

File: napari/utils/theme.py
Comment (tlambert03) [inline napari/utils/theme.py:55]: If syntax style must be one of the supported themes from pygments, then we should either use a string validator here or make it an enum of acceptable values.

File: napari/plugins/hook_specifications.py
Comment (tlambert03) [inline napari/plugins/hook_specifications.py:508]: since `folder` is just an alias for `name`, we probably don't need the additional dict here, right? (seems like the key is always just the `folder` string... which we can call `name`.  then this can just be `Sequence[Dict[str, Union[str, Tuple, List]]]`

File: napari/plugins/hook_specifications.py
Comment (tlambert03) [inline napari/plugins/hook_specifications.py:548]: Please add example here as well since this goes in the docs on the webpage... 
Also please add a new `.. autofunction:: napari_provide_theme` entry somewhere in `docs/plugins/hook_specifications.rst` (That's still not done automatically)

File: napari/utils/theme.py
Comment (tlambert03) [inline napari/utils/theme.py]: I'm for this.... will we uncomment?

File: examples/new_theme.py
Comment (tlambert03) [inline examples/new_theme.py]: will registering a new theme always mean calling both `register_theme()`, followed by `register_napari_theme()`?  Can we call `register_napari_theme` internally whenever a new theme has been registered?
If not, and this needs to be public, it should be exposed in `napari.qt`

File: examples/plugins/theme_plugin.py
Comment (tlambert03) [inline examples/plugins/theme_plugin.py]: this whole example can probably just go in the docstring of the hook_specification, where it will show up on the web docs

File: napari/utils/settings/_defaults.py
Comment (tlambert03) [inline napari/utils/settings/_defaults.py]: This can come in another PR, but I wonder whether we would have designed the `Appearance.theme` field as a `str` here if we had a pydantic model in the first place.  Now it feels kinda funny to have a `class Theme(str)` above instead of just `theme: utils.theme.Theme`  @sofroniewn ?

File: napari/plugins/_plugin_manager.py
Comment (tlambert03) [inline napari/plugins/_plugin_manager.py]: Can you comment on why we're storing the theme data instead of theme instances?
Also, it kinda feels like we're duplicating the validation that Pydantic already does for us here.  how about

```py
try:
    theme = Theme.pars_obj(data)
except Exception as e:  # or pydantic.ValidationError
    warn(f'skipping bad theme: {data}')
```

File: napari/utils/theme.py
Comment (tlambert03) [inline napari/utils/theme.py]: not needed, pydantic provides `list(Theme.__fields__)`.  But also, see my comment down in plugin manager, where this seems to be used.  I would just let pydantic do the validation for us (that's the big win of using a pydantic model for the theme).  No need to check the keys and values and all that stuff ourself

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]: discover_themes() calls `hook.napari_provide_theme.call_historic` (which only needs to be called once) but this is connected to an event that will trigger this every time a new plugin is registered... is the historic hook still necessary?  (a historic hook is usually only ever called once, and you provide a callback here, `register_theme_colors` that will be called _for you_ each time the plugin manager detects a new plugin that provides the hook)

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py:423]: other than triggering discovery when we show the first window, it's not immediately clear to me why all of this logic now lives on the mainwindow.    I feel like we're mixing business logic from the plugin_manager and the settings altogether on the main window here.  most of the new theme-related methods aren't really using the `self` attribute (except to call `_rebuild_theme_settings`, which is just a static method that perhaps belongs on the settings manager instead?)

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py:469]: fyi, I believe you can connect directly to `theme.events.connect` so :`theme.events.connect(lambda _: self._update_theme())` gets them all done in one go  (or at least I think 🤔)

**Review by tlambert03 — COMMENTED** [review]
Pretty awesome stuff @lukasz-migas!

I left a lot of comments... hope you don't mind.

a couple general thoughts that don't really go anywhere specific: 

I think it's pretty cool that we can dynamically change themes now... but do you see this eventually being something that persists across sessions?  If so, it feels like "dark theme" doesn't really mean much anymore, since it is whatever you make it.  If not, it feels significantly less powerful.  So perhaps we need the provided themes to be frozen, and once you start mucking about, you're creating a "custom" theme, but not actually changing the frozen themes in `_themes`

this can come in a follow up PR, but I feel like we're not using the Pydantic model to it's fullest.  With this we now have even more "theme aware" parts.  The plugin manager gains a few themes, settings have some new theme stuff.  The main window has a lot more theme methods... so I generally feel like it could be a little better compartmentalized. 

anyway, let me know if you have thoughts/questions on my comments

**Comment by lukasz-migas** [comment]
@tlambert03 thanks a lot for all the helpful comments and suggestions. Absolutely not a problem😊

I haven't tried addressing the persistence of the theme metadata but I think eventually it would be quite useful to that information in the settings file. I must admit I haven't looked much through issues/PRs to see if this discussion already happened and whether it was something that was to be avoided.

With respect to the loss of meaning for the two default themes, I can see that argument and it would be easy to lock them so users couldn't change them.

I will address your individual comments a bit later but indeed, I've added quite a few extra methods to several places. I think I can scale it down somewhat and keep it to the minimum. The reason for putting quite a few methods on the main window was because that's where the stylesheet are being set. You are right to point out that a lot of these methods can be static so they could live somewhere else.

File: napari/utils/theme.py
Comment (lukasz-migas) [reply napari/utils/theme.py]: I agree. The `folder` attribute could be renamed to `name`. Now would be a good time to do it since we are adding a completely new model to handle theme data.

File: napari/utils/theme.py
Comment (lukasz-migas) [reply napari/utils/theme.py:55]: I've added a value validator that will check against `pygments.styles.STYLE_MAP`. I don't think `pygments` is a direct dependency so this might have to optional?

File: napari/plugins/hook_specifications.py
Comment (lukasz-migas) [reply napari/plugins/hook_specifications.py:508]: Thanks for poinint this out. I've changed to return a sequence of dictionaries.

File: napari/utils/theme.py
Comment (lukasz-migas) [reply napari/utils/theme.py]: For the time being, I've left it commented out since a lot of tests will fail with this warning.

File: napari/plugins/_plugin_manager.py
Comment (lukasz-migas) [reply napari/plugins/_plugin_manager.py]: You are right, I've changed this so `pydantic` does all of the necessary validation.

File: napari/_qt/qt_main_window.py
Comment (lukasz-migas) [reply napari/_qt/qt_main_window.py]: This was actually not necessary so I've removed it. Thanks!

File: napari/_qt/qt_main_window.py
Comment (lukasz-migas) [reply napari/_qt/qt_main_window.py:423]: I've removed a couple of the methods since they were not really required. 

The remaining methods such as `_setup_existing_themes`, `_add_theme`, `_remove_theme`, `_connect_theme` and `_theme_icon_changed` methods are added in the `MainWindow` so any changes to the theme values are reflected in the GUI.

File: napari/_qt/qt_main_window.py
Comment (lukasz-migas) [reply napari/_qt/qt_main_window.py:469]: That's true. I did think of doing it like this and we still can. I've decided against it because the `icon` attribute is a special case where the `_update_theme` method would do nothing.

File: napari/utils/theme.py
Comment (tlambert03) [reply napari/utils/theme.py:55]: > I don't think pygments is a direct dependency so this might have to optional?

That's true, but the only reason we have this option in the theme is basically to support `napari-console`, which depends on qtconsole, and _that_ depends on `pygments` (and if it stopped depending on pygments, then we'd basically need to come up with new themeing anyway).  So, I'd say we should either remove this from our theme entirely (and leave it to `napari-console` to figure out how to theme itself)... or we should just use pygments for now and leave that decision for when we don't require `napari-console`.  If we keep it, here's one option for validating against a set of string literals:

```py
from typing_extensions import Literal as L
import pygments

class Theme(EventedModel):
    s: Union[tuple(L[x] for x in pygments.styles.STYLE_MAP)]
```

File: examples/new_theme.py
Comment (sofroniewn) [reply examples/new_theme.py]: Agreed, it would be nice if we could call `register_napari_theme` internally and this did not need to be exposed publicly

File: examples/plugins/theme_plugin.py
Comment (sofroniewn) [reply examples/plugins/theme_plugin.py]: I'm ok with having it in both places, I sort of like the runnable example in a plugins folder, but example should be in doc string too probably

File: napari/_vispy/vendored/volume.py
Comment (sofroniewn) [inline napari/_vispy/vendored/volume.py]: Can you removed the changed to this vendored file. It will likely cause merge conflicts with some other PRs in progress and they are only formatting changes

File: napari/utils/misc.py
Comment (sofroniewn) [inline napari/utils/misc.py]: Can you add a "parameters" section to the doc string here

File: napari/utils/misc.py
Comment (sofroniewn) [inline napari/utils/misc.py:558]: Can you add a "parameters" section to the doc string here

File: napari/utils/misc.py
Comment (sofroniewn) [inline napari/utils/misc.py:482]: Can you add a "parameters" section to the doc string here. (I know it was already there without one, but now is a good time to add it, thanks!!)

File: napari/utils/settings/_defaults.py
Comment (sofroniewn) [reply napari/utils/settings/_defaults.py]: Would this then allow someone in the settings to immediately create custom themes where they changed individual colors etc. That would be quite nice, but let's leave it for its own PR

File: napari/utils/theme.py
Comment (sofroniewn) [reply napari/utils/theme.py]: Let's rename to "name" in this PR. But should we then reuse this "name" inside our dict of themes, i.e. should this now be the name you use when registering and getting a theme too, or can they be different?

File: napari/utils/theme.py
Comment (sofroniewn) [reply napari/utils/theme.py:55]: > or we should just use pygments for now and leave that decision for when we don't require napari-console

Maybe we do this for now and then refactor when we pull out napari-console? Just seems a little easier, but we can also deal with now

File: napari/utils/theme.py
Comment (sofroniewn) [inline napari/utils/theme.py]: See comment above - but if we rename `folder` to `name`, is it weird if we then have a them with a different name in the `EventedDict` and in the `Theme` itself? Might be nice to force these to be the same somehow. Thoughts @tlambert03?

File: napari/utils/theme.py
Comment (sofroniewn) [reply napari/utils/theme.py]: Agreed we should probably uncomment. We could add to our test skips here https://github.com/napari/napari/blob/115d3ef468d4babb8d722513a1f97002811394a8/pyproject.toml#L84-L96 temporarily and then fix tests in follow up PRs, or fix tests still in this PR

**Review by sofroniewn — COMMENTED** [review]
Looks like this is coming along nicely @lukasz-migas - I just did another review, flagged some questions, mostly minor. Hopefully we should be able to get this in soon!!

File: napari/utils/theme.py
Comment (lukasz-migas) [reply napari/utils/theme.py:55]: Agree, I am happy to change this when napari-console is no longer a dependency in the future.

File: napari/utils/theme.py
Comment (lukasz-migas) [reply napari/utils/theme.py]: Done 👍

File: napari/utils/misc.py
Comment (lukasz-migas) [reply napari/utils/misc.py:558]: Done

File: napari/utils/theme.py
Comment (lukasz-migas) [reply napari/utils/theme.py]: I actually torn on this because I can envisage a situation where a theme is just a variation of an existing theme e.g. `dark` so the icon color is the same and there should be no need to create a new version of the icons under a different  namespace.

File: examples/new_theme.py
Comment (lukasz-migas) [reply examples/new_theme.py]: This is now done internally whenever a new theme is registered in the `_theme` dict so the user doesn't need to call it.

File: napari/utils/theme.py
Comment (sofroniewn) [reply napari/utils/theme.py]: ok, that's fine

**Comment by tlambert03** [comment]
thanks for fixing conflicts @lukasz-migas ... sorry about that.  Is this just waiting for re-review at this point?

**Comment by lukasz-migas** [comment]
> 
> 
> thanks for fixing conflicts @lukasz-migas ... sorry about that. Is this just waiting for re-review at this point?

No problem. It would seem my black auto-formatting is a bit overzelous sometimes...

Yes, this should be ready for another review. I will add a couple extra tests but the rest should be good.

File: examples/cursor_ray.py
Comment (tlambert03) [inline examples/cursor_ray.py]: my black also corrects these... but this bothers my sense of PR granularity :joy:  (i.e. when we look back to see what PR modified a file, if we're going to do a pure formatting change on an unrelated file, I'd like the PR to be something like "black reformat", not "add hook spec")

can you revert this file, and other un-related file formats (like ndLabels... not sure if there are others)?

do you have a `--all-files` setting in your `pre-commit` or black settings somehow?  I'm curious why this would have been modified at all?

File: examples/plugins/theme_plugin.py
Comment (tlambert03) [reply examples/plugins/theme_plugin.py]: but, it's not exactly "runnable" that easily.  It _can_ be runnable, but only if someone uses `napari  examples/plugins/theme_plugin.py` (but just running `python examples/...` would do nothing).  That would be a first for our examples directory.  So at the very least, it should be made clear that you can't actually run this example unless you start it "just so"

**Review by tlambert03 — COMMENTED** [review]
@lukasz-migas 

can you undo all of the formatting-only changes that don't have anything to do with the context of this PR?  makes it a bit harder to review...

File: examples/cursor_ray.py
Comment (lukasz-migas) [reply examples/cursor_ray.py]: I am using the pre-commit config without any modifications. I will check which files were just formatted and undo them.

**Comment by lukasz-migas** [comment]
@tlambert03 Sorry for the auto-formatting issue. I think its because I've had Black file-watcher enabled in Pycharm. 

All auto-formatted has been undone. I've left two files that are not necessarily related to this PR but fix spelling.

File: examples/plugins/theme_plugin.py
Comment (sofroniewn) [reply examples/plugins/theme_plugin.py]: ok good point, maybe we should just move to doc string for now

**Comment by sofroniewn** [comment]
hi @lukasz-migas, sorry this now has some merge conflicts. I think it was looking pretty close though.

We might want to make it `napari_experimental_provide_theme` though as i think @tlambert03 has some ideas for a more declarative syntax in the future that might change this. Hopefully we can get your PR merged soon though!!

**Comment by jni** [comment]
Oh man the gif was hilarious, like a slow motion train wreck, going from a rather pleasant blue background to the fluoro pink at the end. 😂 

I agree that this should be renamed to contain `experimental` in the name, but otherwise is awesome! 👍

**Comment by lukasz-migas** [comment]
> 
> 
> hi @lukasz-migas, sorry this now has some merge conflicts. I think it was looking pretty close though.
> 
> We might want to make it `napari_experimental_provide_theme` though as i think @tlambert03 has some ideas for a more declarative syntax in the future that might change this. Hopefully we can get your PR merged soon though!!

No problem, I will resolve the conflicts later today and also change the name to be considered experimental 😄

File: napari/utils/theme.py
Comment (tlambert03) [inline napari/utils/theme.py:191]: quick note here, with this warning, everyone will be seeing a warning on every napari usage now because of napari-console:
```
/Users/talley/miniconda3/envs/napdev/lib/python3.9/site-packages/napari_console/qt_console.py:161: FutureWarning: Themes were changed to use evented model with Pydantic's color type rather than the `rgb(x, y, z)`. You can get the old color by calling `color.as_rgb()`. The `as_dict=True` option will be removed in 0.X.X
  theme = get_theme(self.viewer.theme)
```

Comment (lukasz-migas) [reply napari/utils/theme.py:191]: That's true - its can be very easily fixed in the `napari-console` plugin but until then it will be showing each time. 

The startup warning is actually caused by the fact that I am connecting the `theme.console` and `theme.syntax_style` events directly to the `qt_viewer.console._update_theme` methods. I could wrap with `warnings.catch_warnings` for the time being?

Comment (tlambert03) [reply napari/utils/theme.py:191]: I do think it's pretty jarring to have everyone see a warning in our default install.  So either here, or in a follow-up PR before we release, we should either do the catch_warnings... or we should update napari-console and pin our dependency here

**Comment by sofroniewn** [comment]
Unfortunately we've still got 6 failing tests here - the `min_req` test is failing a lot on other PRs right now, but this is the only PR with so many fails. If we could get it down to just 1 or 2 I'd feel better about merging this. Maybe @jni or @tlambert03 might have ideas on why some of those extra fails are happening. Sorry this has been such a long PR process, will be great to get this in once more tests are passing!!

**Comment by lukasz-migas** [comment]
@sofroniewn thanks for re-running them. 

I just had a look at the output and the four ubuntu tests are not actually failing at the test stage but rather after. They seem to be failing with `XIO:  fatal IO error 0 (Success) on X server ":0"` which might be related pytest-xvfb.

There seems to be a segfault on Windows though...

**Comment by tlambert03** [comment]
I'm having a hard time getting any tests to pass without segfault when I remote into a linux box (even on master)

**Comment by sofroniewn** [comment]
> I'm having a hard time getting any tests to pass without segfault when I remote into a linux box (even on master)

@tlambert03 if you just try running the tests that are added in this PR or just try skipping the tests added in this PR can you narrow it down to a few that we might be able to skip on CI? We could then merge the code which is working and unskip the tests on CI once we figure more things out.

Just trying to think about how we can unblock this and get it merged as I know a lot of work has gone into this one!!

**Comment by sofroniewn** [comment]
@Czaki if you could take a look at test fails here too it would be greatly appreciated, thanks!!

**Comment by tlambert03** [comment]
hey @lukasz-migas, I see you're disabling some things to test.  i've been working on debugging tests yesterday and today. I've narrowed it down to the perfmon test in most cases. I might have some fixes later today (though feel free to keep exploring yourself too!)

**Comment by lukasz-migas** [comment]
@tlambert03 yes, it's been bugging me for days to figure out what the issue is with this PR but I haven't had time to actually do it. Thanks for taking a look yourself! Look forward to seeing what the problem is...

**Comment by tlambert03** [comment]
@lukasz-migas, you could try merging #3167 into this PR to see how it does.  We should ultimately merge it independently/before this ... but if you're curious to see if it will work before people have time to review/merge the other one you could try it

**Comment by sofroniewn** [comment]
I have optimistically merged #3167, so you should just be able to update from master

**Comment by tlambert03** [comment]
ok, looks like there are still PR-specific issues here (on windows at least).  @lukasz-migas ... I'd start by commenting out all of the new tests you added here, to see if the old tests still work in isolation with the new code (i.e. whether it's a new test side effect or a code problem)

**Comment by Czaki** [comment]
```
     File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\concurrent\futures\thread.py", line 78 in _worker
    File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\threading.py", line 870 in run
    File "D:\a\napari\napari\napari\utils\notifications.py", line 319 in run_with_except_hook
    File "D:\a\napari\napari\napari\utils\notifications.py", line 319 in run_with_except_hook
    File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\threading.py", line 926 in _bootstrap_inner
    File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\threading.py", line 890 in _bootstrap
```

Thi looks like a problem with notification manager. I found that where is workaround for thread excepthook apply in `install_hook` (line 222)

https://github.com/napari/napari/blob/2381c71b942492cfd8f2d402a803aca54a14337e/napari/utils/notifications.py#L210-L228

 but this workaround was not cleaned during restore_hooks. 

https://github.com/napari/napari/blob/2381c71b942492cfd8f2d402a803aca54a14337e/napari/utils/notifications.py#L230-L239

So maybe add proper cleaning step is a good idea?

**Comment by sofroniewn** [comment]
So funny the remaining windows test fails are out of memory errors. I've not seen them on other PRs, really not sure why they are occurring here. I think we're close though to having this all figured out!!

```
  napari\_vispy\_tests\test_vispy_big_images.py ...FEWindows fatal exception: code 0xc0000008
```

**Comment by lukasz-migas** [comment]
@sofroniewn I currently disabled a lot of the tests to see if that made a difference. It did to the non-windows os's. I will re-enable them tomorrow to see if they still pass after changes to xvfb.

I do hope we are getting close 😱😱😱

**Comment by sofroniewn** [comment]
From Zulip:
> I am sorry I haven't had chance to work on this the last week. I am away on holiday cut off from my PC and not back until next Wednesday.

As mentioned there all good! Enjoy your time off!!!

> I do hope we are getting close 😱😱😱

It does look like the windows tests are still giving us problems. One idea for moving forward is if you could pull out the EventedTheme stuff into its own PR and leave the hookspec/ plugin stuff in this one. That EventedTheme PR should hopefully be more straight forward to merge with no failing tests and then we can revisit and debug the hookspec/ plugin stuff more easily.

I know you've put so much time into this already, sorry that we still keep running up into problems, but splitting the PR into smaller chunks will hopefully make the merge process easier and let us get the great functionality in sooner. Our apologies for not thinking of this sooner.

**Comment by lukasz-migas** [comment]
image

PkOc

Yay, it finally passed all tests...

@sofroniewn Thanks for the suggestion of splitting the `Evented Theme` to another PR but I don't think that would have helped. I am pretty sure the issue here was the `Theme` model (and all code associated with it) and not the actual hook spec. After many attempts and @tlambert03 kind fixes for the `pytest-xvfb` I think the issue was with the the way the teardown/cleanup method was taking place. The `Window` instance implements `close` method which removes the `_qt_window` when its called and I think that might have been the problem. In any case, I've added a new method to explictly clean-up and disconnect events connected to `_themes` instance and each `Theme` which seems to solve the issue.

**Review by sofroniewn — APPROVED** [review]
They are all passing!!

Heroic work here and well figured out with the teardown. I'd like to merge this now so that it gets into the next release candidate, but I'll wait for @tlambert03 to approve too.

Thanks again, and onwards and upwards! 🎆 🚀

**Comment by sofroniewn** [comment]
Ok, I'm going to go with merging this now and getting it in. We'll still do one more round of RC testing. It's flagged as experimental, and all tests are passing, so let's give it a go!!

**Pull request merged**
